sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -575,15 +575,43 @@ case class WrapOption(child: Expression, optType: DataType)
   }
 }
 
+object LambdaVariable {
+  private val curId = new java.util.concurrent.atomic.AtomicLong()
+
+  // Returns the codegen-ed `LambdaVariable` and add it to mutable states, so that it can be
+  // accessed anywhere in the generated code.
+  def prepareLambdaVariable(ctx: CodegenContext, variable: LambdaVariable): ExprCode = {
+    val variableCode = variable.genCode(ctx)
+    assert(variableCode.code.isEmpty)
+
+    ctx.addMutableState(
+      CodeGenerator.javaType(variable.dataType),
+      variableCode.value,
+      forceInline = true,
+      useFreshName = false)
+
+    if (variable.nullable) {
+      ctx.addMutableState(
+        CodeGenerator.JAVA_BOOLEAN,
+        variableCode.isNull,
+        forceInline = true,
+        useFreshName = false)
+    }
+
+    variableCode
+  }
+}
+
 /**
- * A placeholder for the loop variable used in [[MapObjects]].  This should never be constructed
+ * A placeholder for the loop variable used in [[MapObjects]]. This should never be constructed
  * manually, but will instead be passed into the provided lambda function.
  */
+// TODO: Merge this and `NamedLambdaVariable`.
 case class LambdaVariable(
-    value: String,
-    isNull: String,
+    name: String,
     dataType: DataType,
-    nullable: Boolean = true) extends LeafExpression with NonSQLExpression {
+    nullable: Boolean,
+    id: Long = LambdaVariable.curId.incrementAndGet) extends LeafExpression with NonSQLExpression {
```

Comment:
Probably add comment about the `id`? The allocated `id`s here can't be directly used. It needs going through `ReassignLambdaVariableID` to reassign. If the expressions including `LambdaVariable` skip the normal query processing steps, we still make sure `ReassignLambdaVariableID` is applied.

Reply:
> The allocated ids here can't be directly used

Yes they can. A globally unique ID is fine here, it just breaks the codegen cache.

Reply:
There we do `abs` on positive and negative ids, if globally unique id is used, won't it probably get conflict?

Reply:
see the newly added check. If `ReassignLambdaVariableID` is applied, then all IDs are negative and unique. If the rule is not applied, then all IDs are positive and unique.

Reply:
AFAIK this is safe as we never combine optimized plans.